### ISSUE TITLE
[Scala] Further Enhancements

### ISSUE THREAD

**Issue by djspiewak** [sublimehq/Packages#588]
There are some further enhancements which are included in #94 which should be ported into the mainline Scala mode.  Off the top of my head:

- [ ] Unicode character escapes in strings (e.g. `\u2101`)
- [ ] Scoping for annotations (outside of primitives like `@volatile`)
- [ ] Better handling for XML
- [ ] Meta scoping
- [ ] Sane scaladoc highlighting

I understand there are several other features also implemented by #94, but I'm personally of the *opinion* that they're probably not worth pulling in (e.g. specifically target-scoped braces).

This is the place for opinions though!  Ping @gwenzek for thoughts, opinions and insights.  Anything I missed?  Anything that should be added to the above?

**Comment by gwenzek** [comment]
Thanks for making this list.

I'd also like to add:

- [ ] Scoping of function call `f(a)`
- [ ] Scoping of operators in infix position `a + b`
- [ ] Similar scopes for `s"$foo"` and `s"${foo}"`

Regarding the symbol list I'd like to improve it with the following points:

- [ ] Full function signature in  symbol list
- [ ] Indentation in symbol list

I think the point you mentioned are the most consensual, so I'll start by that.

**Comment by gwenzek** [comment]
Below are two screenshots showing what the symbol list could look like (I don't have `val` here but it's just an example).

**Current symbol list**:
![Current symbol list](https://cloud.githubusercontent.com/assets/5920036/17756132/5abcfafe-64de-11e6-96cc-918c17570f52.png)

**Proposed symbol list**
![Proposed symbol list](https://cloud.githubusercontent.com/assets/5920036/17755981/78d16288-64dd-11e6-8fbd-9b93bc8bca08.png)

IMHO the second one is more readable. The problem comes from too long signatures that are strangely abbreviated, instead of just truncated.

**Comment by djspiewak** [comment]
> - Full function signature in symbol list
> - Indentation in symbol list

Yes yes yes!  I took a crack at doing this a few weeks ago.  This is one thing that the Python mode does tremendously well.

Unfortunately, Scala's function and value declaration syntax is context-free, whereas Python's is regular, which means that we're not going to make things quite as good as we might like.

> Below are two screenshots showing what the symbol list could look like

I would be in favor of removing the `trait`, `class`, `def` stuff, as well as likely the parameter names.  As I said, the Python mode does a great job here, but I'm not sure how far in that direction we can go.

**Comment by djspiewak** [comment]
> Scoping of function call `f(a)`

I'm definitely not in favor of this, simply because the following two lines will be scoped differently despite being identical in semantics:

```scala
x map f
x.map(f)
```

> Scoping of operators in infix position `a + b`

I'm opposed to this one as well, for basically the same reason.  I don't want the following two lines to be scoped differently:

```scala
a + b
a map b
```

> Similar scopes for `s"$foo"` and `s"${foo}"`

I agree on this one, but I'm skeptical about how it's possible.  Are you just going to special case `\$\{\s*{{id}}\s*\}`?  Or are you proposing to change the scoping on `$foo`?

**Comment by wbond** [comment]
- [ ] `punctuation.*` and `meta.*` scopes for `{`, `}`, `(`, `)`, `[`, `]`
